File: 64.最小路径和.go
/*
 * @lc app=leetcode.cn id=64 lang=golang
 *
 * [64] 最小路径和
 */
package leetcode

// @lc code=start
func minPathSum(grid [][]int) int {
	min := func(a, b int) int {
		if a < b {
			return a
		}
		return b
	}

	m, n := len(grid), len(grid[0])

	dp := make([][]int, m)
	for i := 0; i < m; i++ {
		dp[i] = make([]int, n)
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if i == 0 && j == 0 {
				dp[i][j] = grid[i][j]
			} else if i == 0 {
				dp[i][j] = dp[i][j-1] + grid[i][j]
			} else if j == 0 {
				dp[i][j] = dp[i-1][j] + grid[i][j]
			} else {
				dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j]
			}

		}
	}

	return dp[m-1][n-1]
}

// @lc code=end

File: rust/64.最小路径和.rs
/*
 * @lc app=leetcode.cn id=64 lang=rust
 *
 * [64] 最小路径和
 */

// @lc code=start
impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        let m = grid.len();
        let n = grid[0].len();
        let mut dp: Vec<Vec<i32>> = vec![vec![grid[0][0];n];m];
        for i in 1..m {
            dp[i][0] = grid[i][0] + dp[i-1][0];
        }
        for j in 1..n {
            dp[0][j] = grid[0][j] + dp[0][j-1];
        }
        for i in 1..m {
            for j in 1..n {
                let a = dp[i-1][j];
                let b = dp[i][j-1];
                dp[i][j] = grid[i][j] + {if a < b {a} else {b}};
            }
        }
        return dp[m-1][n-1];
    }
}
// @lc code=end

